src/common/hooks/use-transaction-status.ts

```diff
@@ -0,0 +1,34 @@
+import React, { useEffect, useState } from 'react';
+import { MempoolTransaction, Transaction } from '@stacks/stacks-blockchain-api-types';
+import { TxStatus } from '@common/types/tx';
+import { TransactionStatus } from '@common/constants';
+
+export const useTransactionStatus = (
```

Comment:
Does this need to be a hook? It isn't reliant on any external state. So it can just be a function that accepts a transaction and returns a value.

Reply:
I don't think using a custom hook is dependent on relying on external state? It is extracting out setting state into a reusable hook? I did forget the dependency array though. We seem to use custom hooks in a lot of places that don't rely on external state. Not right?

Reply:
Sorry for essay, hope this is a helpful explanation.

You're right, hooks aren't only reliant on using external state, but they enable access to a handful of features right. Access to state methods that persist between renders, `useEffect` to manage code dependent on changing props, other hooks etc.

It's not clear, based on this implementation, why we need any of these. The `useState` is updated on every change, but why not just return the value immediately?

Consider if we need to reuse this code elsewhere: directly in an atom, within another function etc. We're unable to because of the decision to make it a hook, as they only work in components/other hooks. 
Sometimes we need hooks of course, for one of the reasons stated above, but the core logic of it is usually better described as a pure function. 

Here's a simplified implementation:

```ts
export function getTransactionStatus(tx: Transaction | MempoolTransaction) {
  if (tx?.tx_status === 'pending') {
    return TransactionStatus.PENDING;
  } else if (tx?.tx_status === 'success' && tx.is_unanchored) { // <-- note, no need for `!!` here, as the value is already a boolean
    return TransactionStatus.SUCCESS_MICROBLOCK;
  } else if (tx?.tx_status === 'success' && !tx.is_unanchored) {
    if (!tx.microblock_canonical) {
      return TransactionStatus.FAILED;
    } else {
      return TransactionStatus.SUCCESS_ANCHOR_BLOCK;
    }
  } else if (tx?.tx_status === 'abort_by_response' || tx?.tx_status === 'abort_by_post_condition') {
    return TransactionStatus.FAILED;
  }
}
```

Another important change I've made is to not bake in the `if undefined, return undefined` logic. This is additional functionality that 1) breaks the single responsibility principle, 2) makes an assumption the tx can sometimes be `undefined`. See it as a coupling of getting the tx status _and_ allowing it to be undefined. It's almost always better to leave the handling of this to whatever code invokes the function, rather than the fn itself. 

It also leverages the TypeScript compiler. If I use this again, and I pass it a value that might be undefined, the compiler will tell me. It's a silent, unobvious feature if it accepts `undefined`.

If you want to get #functional. We could make a wrapper method that composes this functionality. (Though this isn't necessary in this example).

```js
 // untyped example
function acceptUndefined (fn) {
  return (value) => value === undefined ? undefined : fn(value);
}
const newFunctionThatAcceptUndefined = acceptUndefined(getTransactionStatus)
```

Reply:
I appreciate the explanation. I def get your point now, thanks for taking the time to write it up!